File: java-bridge-matlab/src/main/java/idesyde/matlab/SimulinkReactiveDesignModel.java
package idesyde.matlab;

import idesyde.core.DesignModel;
import idesyde.core.headers.DesignModelHeader;
import idesyde.core.headers.LabelledArcWithPorts;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public record SimulinkReactiveDesignModel(
        Set<String> processes,
        Map<String, Long> processesSizes,
        Set<String> delays,
        Map<String, Long> delaysSizes,
        Set<String> sources,
        Map<String, Long> sourcesSizes,
        Map<String, Double> sourcesPeriods,
        Set<String> constants,
        Set<String> sinks,
        Map<String, Long> sinksSizes,
        Map<String, Double> sinksDeadlines,
        Map<String, Map<String, Map<String, Long>>> processesOperations,
        Map<String, Map<String, Map<String, Long>>> delaysOperations,
        Set<SimulinkArc> links
) implements DesignModel {
    @Override
    public DesignModelHeader header() {
        var elems = new HashSet<String>();
        elems.addAll(processes);
        elems.addAll(delays);
        elems.addAll(sources);
        elems.addAll(constants);
        elems.addAll(sinks);
        return new DesignModelHeader(
                "SimulinkReactiveDesignModel",
                elems,
                links.stream().map(l ->
                        new LabelledArcWithPorts(l.src(), l.dst(), l.srcPort(), l.dstPort(), String.valueOf(l.dataSize()))).collect(Collectors.toSet()
                ),
                new HashSet<>()
        );
    }
}

File: kotlin-bridge-matlab/src/main/kotlin/idesyde/matlab/SimulinkReactiveDesignModel.kt
package idesyde.matlab

import idesyde.core.DesignModel
import idesyde.core.headers.DesignModelHeader
import idesyde.core.headers.LabelledArcWithPorts

data class SimulinkReactiveDesignModel(
    val processes: Set<String>,
    val processesSizes: Map<String, Long>,
    val delays: Set<String>,
    val delaysSizes: Map<String, Long>,
    val sources: Set<String>,
    val sourcesSizes: Map<String, Long>,
    val sourcesPeriods: Map<String, Double>,
    val constants: Set<String>,
    val sinks: Set<String>,
    val sinksSizes: Map<String, Long>,
    val sinksDeadlines: Map<String, Double>,
    val processesOperations: Map<String, Map<String, Map<String, Long>>>,
    val delaysOperations: Map<String, Map<String, Map<String, Long>>>,
    val links: Set<SimulinkArc>
) : DesignModel {
    override fun header(): DesignModelHeader =
        DesignModelHeader(
            category = "SimulinkReactiveDesignModel",
            elements = (processes + delays + sources + constants + sinks).toList(),
            modelPaths = listOf(),
            relations = links.map { l ->
                LabelledArcWithPorts(
                    src = l.src,
                    dst = l.dst,
                    srcPort = l.srcPort,
                    dstPort = l.dstPort,
                    label = l.dataSize.toString()
                )
            }

        )
}

